### PR TITLE
Build against LLVM 16.06

### PR DESCRIPTION
**Pull request by certik**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by certik** [comment]
@conda-forge-admin, please rerender

**Comment by certik** [comment]
@conda-forge-admin, please rerender

**Comment by certik** [comment]
It failed with
```
Undefined symbols for architecture x86_64:
  "_clonefile", referenced from:
      llvm::sys::fs::copy_file(llvm::Twine const&, llvm::Twine const&) in libLLVMSupport.a(Path.cpp.o)
  "_pthread_set_qos_class_self_np", referenced from:
      llvm::set_thread_priority(llvm::ThreadPriority) in libLLVMSupport.a(Threading.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```
But these errors I believe can be fixed by updating MACOSX_SDK_VERSION.

**Comment by certik** [comment]
osx/x64 works. osx/arm works for all previous bugs, but has a new one:
```
[ 78%] Linking CXX executable lfortran
ld: warning: ignoring file /Users/runner/miniforge3/conda-bld/lfortran_1697098221494/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/libzstd.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
Undefined symbols for architecture x86_64:
  "_ZSTD_compress", referenced from:
      llvm::compression::compress(llvm::compression::Params, llvm::ArrayRef<unsigned char>, llvm::SmallVectorImpl<unsigned char>&) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::zstd::compress(llvm::ArrayRef<unsigned char>, llvm::SmallVectorImpl<unsigned char>&, int) in libLLVMSupport.a(Compression.cpp.o)
  "_ZSTD_compressBound", referenced from:
      llvm::compression::compress(llvm::compression::Params, llvm::ArrayRef<unsigned char>, llvm::SmallVectorImpl<unsigned char>&) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::zstd::compress(llvm::ArrayRef<unsigned char>, llvm::SmallVectorImpl<unsigned char>&, int) in libLLVMSupport.a(Compression.cpp.o)
  "_ZSTD_decompress", referenced from:
      llvm::compression::decompress(llvm::DebugCompressionType, llvm::ArrayRef<unsigned char>, unsigned char*, unsigned long) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::zstd::decompress(llvm::ArrayRef<unsigned char>, unsigned char*, unsigned long&) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::zstd::decompress(llvm::ArrayRef<unsigned char>, llvm::SmallVectorImpl<unsigned char>&, unsigned long) in libLLVMSupport.a(Compression.cpp.o)
  "_ZSTD_getErrorName", referenced from:
      llvm::compression::decompress(llvm::DebugCompressionType, llvm::ArrayRef<unsigned char>, unsigned char*, unsigned long) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::zstd::decompress(llvm::ArrayRef<unsigned char>, unsigned char*, unsigned long&) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::zstd::decompress(llvm::ArrayRef<unsigned char>, llvm::SmallVectorImpl<unsigned char>&, unsigned long) in libLLVMSupport.a(Compression.cpp.o)
  "_ZSTD_isError", referenced from:
      llvm::compression::compress(llvm::compression::Params, llvm::ArrayRef<unsigned char>, llvm::SmallVectorImpl<unsigned char>&) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::zstd::compress(llvm::ArrayRef<unsigned char>, llvm::SmallVectorImpl<unsigned char>&, int) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::decompress(llvm::DebugCompressionType, llvm::ArrayRef<unsigned char>, unsigned char*, unsigned long) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::zstd::decompress(llvm::ArrayRef<unsigned char>, unsigned char*, unsigned long&) in libLLVMSupport.a(Compression.cpp.o)
      llvm::compression::zstd::decompress(llvm::ArrayRef<unsigned char>, llvm::SmallVectorImpl<unsigned char>&, unsigned long) in libLLVMSupport.a(Compression.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```
This is due to zstd. I think this issue is unrelated.

**Comment by certik** [comment]
@conda-forge-admin, please rerender

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -1,4 +1,6 @@
+MACOSX_SDK_VERSION:        # [osx and x86_64]
+  - "11.3"                 # [osx and x86_64]
 c_compiler_version:            # [osx]
-  - 15                         # [osx]
+  - 16                         # [osx]
 cxx_compiler_version:          # [osx]
-  - 15                         # [osx]
+  - 16                         # [osx]
```
Comment (Shaikh-Ubaid) [inline recipe/conda_build_config.yaml:6]: It seems by default it picks `clang 16`. So, I think we need not mention`c_compiler_version` and `cxx_compiler_version` here.

Comment (Shaikh-Ubaid) [reply recipe/conda_build_config.yaml:6]: But may be we can mention, so that we use a fixed/pinned version at the CI.

**Review by Shaikh-Ubaid — APPROVED** [review]
This looks great to me! Thank you so much for this @certik! We are now building LFortran with LLVM 16!

File: recipe/meta.yaml [hunk 2]
```diff
@@ -23,16 +23,17 @@ requirements:
...
     - cppzmq            # [build_platform != target_platform]
     - zlib              # [build_platform != target_platform]
+    - zstd-static =1.5.5  # [build_platform != target_platform]
   host:
     - xeus =3.0.5
     - xeus-zmq =1.0.2
-    - llvmdev =11.1.0   # [osx and x86_64]
-    - llvmdev =15.0.7   # [not osx or not x86_64]
+    - llvmdev =16.0.6   # [osx and x86_64]
+    - llvmdev =16.0.6   # [not osx or not x86_64]
```
[the remark this answers is not in the document]
Comment (certik) [reply recipe/meta.yaml:36]: https://github.com/conda-forge/lfortran-feedstock/pull/60

File: recipe/conda_build_config.yaml (hunk 1 above)
Comment (certik) [reply recipe/conda_build_config.yaml:6]: https://github.com/conda-forge/lfortran-feedstock/pull/61

File: recipe/meta.yaml [hunk 3]
```diff
@@ -23,16 +23,17 @@ requirements:
     - make   # [unix]
     - xeus =3.0.5       # [build_platform != target_platform]
     - xeus-zmq =1.0.2   # [build_platform != target_platform]
-    - llvmdev =15.0.7   # [build_platform != target_platform]
+    - llvmdev =16.0.6   # [build_platform != target_platform]
     - xtl               # [build_platform != target_platform]
     - nlohmann_json     # [build_platform != target_platform]
     - cppzmq            # [build_platform != target_platform]
     - zlib              # [build_platform != target_platform]
+    - zstd-static =1.5.5  # [build_platform != target_platform]
```
Comment (isuruf) [inline recipe/meta.yaml:31]: Why use the static library?

Comment (certik) [reply recipe/meta.yaml:31]: I wrote up our reasons here: https://github.com/conda-forge/lfortran-feedstock/pull/48#issuecomment-1753237103.

**Pull request merged**